lib/bookmark_query.rb

```diff
@@ -54,27 +64,97 @@ def list_all
 
     results = results.limit(@limit)
 
+    BookmarkQuery.preload(results, self)
+    results
+  end
+
+  private
+
+  def list_all_results(topics, pms)
+    results = base_bookmarks.merge(topics.or(pms))
+    results = results.merge(Post.secured(@guardian))
+    results = @guardian.filter_allowed_categories(results)
+
     if BookmarkQuery.preloaded_custom_fields.any?
-      Topic.preload_custom_fields(
-        results.map(&:topic), BookmarkQuery.preloaded_custom_fields
-      )
+      Topic.preload_custom_fields(results, BookmarkQuery.preloaded_custom_fields)
     end
 
-    BookmarkQuery.preload(results, self)
-
-    @guardian.filter_allowed_categories(results)
+    results
   end
 
-  private
+  def list_all_results_polymorphic(topics, pms)
+    topic_results = topic_bookmarks.merge(topics.or(pms))
+
+    post_results = post_bookmarks.merge(topics.or(pms))
+    post_results = post_results.merge(Post.secured(@guardian))
+
+    topic_results = @guardian.filter_allowed_categories(topic_results)
+    post_results = @guardian.filter_allowed_categories(post_results)
+
+    if BookmarkQuery.preloaded_custom_fields.any?
+      Topic.preload_custom_fields(topic_results, BookmarkQuery.preloaded_custom_fields)
+      Topic.preload_custom_fields(post_results, BookmarkQuery.preloaded_custom_fields)
+    end
 
-  def user_bookmarks
-    # There is guaranteed to be a TopicUser record if the user has bookmarked
-    # a topic, see BookmarkManager
+    # TODO: At some point we may want to introduce ways for other Bookmarkable types
+    # to further filter results securely using merges, though this is not necessary just
+    # yet.
+
+    results = Bookmark.select("bookmarks.*").from(
+      "(#{topic_results.to_sql} UNION #{post_results.to_sql} UNION #{other_bookmarks.to_sql}) as bookmarks"
```

Comment:
Minor but I wonder if there is a way to avoid running the `otther_bookmarks` SQL query if no addition polymorphic type has been registered.

Reply:
Good point, I will just leave out this whole `UNION` dance if there isn't any...though in saying that Sam did want me to look at everything using the same `register_bookmarkable` interface including topics and posts.

Reply:
I avoided this other_bookmarks addition if there are no additional polymorphic types